### PR TITLE
feat: new final hearing and issue resolution state (FPLA-2043 & FPLA-2044)

### PR DESCRIPTION
**Pull request by Jamesnelson117**
**Before creating a pull request make sure that:**

- [x] PR's title includes JIRA ticket number and follow [Conventional Commits](https://www.conventionalcommits.org/) specification
- [x] README and other documentation has been updated / added (if needed)
- [x] tests have been updated / new tests has been added (if needed)

### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/FPLA-2043
https://tools.hmcts.net/jira/browse/FPLA-2044

### Change description ###
Adds 'Final hearing' and 'Issue resolution' case state
Final hearing case state progressed by review CMO in which next hearing is of Final hearing type
Issue resolution case state progressed by review CMO in which next hearing is of issue resolution type

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ReviewCMOControllerAboutToSubmitTest.java [hunk 1]
```diff
@@ -117,22 +119,33 @@ void shouldSealPDFAndAddToSealedCMOsListWhenJudgeApprovesOrder() throws Exceptio
...
 
+    @Test
+    void shouldUpdateStateToFinalHearingWhenNextHearingIssueIsOfTypeFinalAndCmoDecisionIsSendToAllParties() {
+        given(featureToggleService.isNewCaseStateModelEnabled()).willReturn(true);
+
+        UUID cmoId = UUID.randomUUID();
+        CaseData caseData = buildCaseData(cmoId, SEND_TO_ALL_PARTIES, FINAL);
+        CaseData responseData = extractCaseData(postAboutToSubmitEvent(caseData));
+
+        assertThat(State.FINAL_HEARING).isEqualTo(responseData.getState());
```
Comment (toireasam) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ReviewCMOControllerAboutToSubmitTest.java:140]: Minor: could remove State. and just do FINAL_HEARING

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ReviewCMOControllerAboutToSubmitTest.java [hunk 2]
```diff
@@ -117,22 +119,33 @@ void shouldSealPDFAndAddToSealedCMOsListWhenJudgeApprovesOrder() throws Exceptio
...
-        CaseData caseData = buildCaseData(cmoId, SEND_TO_ALL_PARTIES);
+        CaseData caseData = buildCaseData(cmoId, SEND_TO_ALL_PARTIES, ISSUE_RESOLUTION);
         CaseData responseData = extractCaseData(postAboutToSubmitEvent(caseData));
 
         assertThat(State.ISSUE_RESOLUTION).isEqualTo(responseData.getState());
     }
 
+    @Test
+    void shouldUpdateStateToFinalHearingWhenNextHearingIssueIsOfTypeFinalAndCmoDecisionIsSendToAllParties() {
+        given(featureToggleService.isNewCaseStateModelEnabled()).willReturn(true);
```
Comment (toireasam) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ReviewCMOControllerAboutToSubmitTest.java:134]: thoughts on another test shouldNotUpdateStateToFinalHearingWhenNextHearingIssueIsNotOfTypeFinal ? happy if you think it's covered enough though!

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ReviewCMOControllerAboutToSubmitTest.java [hunk 3]
```diff
@@ -117,22 +119,33 @@ void shouldSealPDFAndAddToSealedCMOsListWhenJudgeApprovesOrder() throws Exceptio
...
+        UUID cmoId = UUID.randomUUID();
+        CaseData caseData = buildCaseData(cmoId, SEND_TO_ALL_PARTIES, FINAL);
+        CaseData responseData = extractCaseData(postAboutToSubmitEvent(caseData));
+
+        assertThat(State.FINAL_HEARING).isEqualTo(responseData.getState());
+    }
+
     @Test
     void shouldNotUpdateStateToIssueResolutionWhenFeatureToggledOff() {
         given(featureToggleService.isNewCaseStateModelEnabled()).willReturn(false);
```
Comment (toireasam) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ReviewCMOControllerAboutToSubmitTest.java:146]: Do we need similar feature toggle test for Final Hearing state?

Comment (toireasam) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ReviewCMOControllerAboutToSubmitTest.java:146]: Or maybe this test name could be changed to doNotUpdateStateWhenFeatureToggleIsOff to cater for both?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/cmo/ReviewCMOController.java [hunk 4]
```diff
@@ -118,9 +119,13 @@ public AboutToStartOrSubmitCallbackResponse handleAboutToSubmit(@RequestBody Cal
                 data.put("sealedCMOs", sealedCMOs);
 
                 if (featureToggleService.isNewCaseStateModelEnabled()
-                    && caseData.getReviewCMODecision().hasReviewOutcomeOf(SEND_TO_ALL_PARTIES)
-                    && isNextHearingOfType(caseData, cmo.getId(), ISSUE_RESOLUTION)) {
-                    data.put("state", State.ISSUE_RESOLUTION.getValue());
+                    && caseData.getReviewCMODecision().hasReviewOutcomeOf(SEND_TO_ALL_PARTIES)) {
+
+                    if (isNextHearingOfType(caseData, cmo.getId(), ISSUE_RESOLUTION)) {
+                        data.put("state", State.ISSUE_RESOLUTION.getValue());
```
Comment (toireasam) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/cmo/ReviewCMOController.java:125]: Again could remove State. and use ISSUE_RESOLUTION if you want - minor

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ReviewCMOControllerAboutToSubmitTest.java (hunk 3 above)
Comment (toireasam) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ReviewCMOControllerAboutToSubmitTest.java:146]: let me know your thoughts :)

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/cmo/ReviewCMOController.java (hunk 4 above)
Comment (Jamesnelson117) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/cmo/ReviewCMOController.java:125]: Yeah I purposely left this extended to State.ISSUE_RESOLUTION to differ between hearing type of same name - line 124. I think it was easier to read at a glance. I then done the same thing with 'State.FINAL_HEARING' just to follow this pattern.

Thoughts?

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ReviewCMOControllerAboutToSubmitTest.java (hunk 3 above)
Comment (Jamesnelson117) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ReviewCMOControllerAboutToSubmitTest.java:146]: hmm, what I think might be better is test - `shouldNotChangeStateWhenNewStateFeatureToggledIsToggledOff` and in that confirm both.

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ReviewCMOControllerAboutToSubmitTest.java (hunk 1 above)
Comment (Jamesnelson117) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ReviewCMOControllerAboutToSubmitTest.java:140]: Could do, please see my reply in last comment.

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ReviewCMOControllerAboutToSubmitTest.java (hunk 2 above)
Comment (Jamesnelson117) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ReviewCMOControllerAboutToSubmitTest.java:134]: I think we can maybe combine the existing.

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ReviewCMOControllerAboutToSubmitTest.java (hunk 3 above)
Comment (Jamesnelson117) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ReviewCMOControllerAboutToSubmitTest.java:146]: Yep I like that idea, done.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/cmo/ReviewCMOController.java (hunk 4 above)
Comment (toireasam) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/cmo/ReviewCMOController.java:125]: Oh yes this makes sense for me.

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ReviewCMOControllerAboutToSubmitTest.java [hunk 5]
```diff
@@ -163,7 +165,7 @@ void shouldNotUpdateStateToIssueResolutionWhenReviewDecisionIsNotSendToAllPartie
     }
 
     @Test
-    void shouldNotUpdateStateToIssueResolutionWhenNextHearingIsNotOfTypeIssueResolution() {
+    void shouldNotUpdateStateToIssueResolutionOrFinalHearingWhenNextHearingIsNotOfTypeIssueResolution() {
```
Comment (toireasam) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ReviewCMOControllerAboutToSubmitTest.java]: maybe rename to shouldNotUpdateStateToIssueResolutionOrFinalHearingWhenNextHearingIsNotOfEitherType ?

**Review by akomisarek — CHANGES_REQUESTED** [review]
Small comments + toggling refactor here.

**Review by akomisarek — DISMISSED** [review]

**Review by akomisarek — APPROVED** [review]

**Pull request merged**